File: algorithms/[289]生命游戏/solution.py
# topics = ["数组"]

from typing import List


class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        0 —> 0  ... 0
        1 —> 1  ... 1
        0 —> 1  ... 2
        1 —> 0  ... 3
        """
        m = len(board)
        n = len(board[0])
        next = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]

        for i in range(m):
            for j in range(n):
                count = 0
                for cx, cy in next:
                    nx, ny = cx + i, cy + j
                    if 0 <= nx < m and 0 <= ny < n:
                        if board[nx][ny] == 1 or board[nx][ny] == 3:
                            count += 1
                if board[i][j] == 0 and count == 3:
                    board[i][j] = 2
                elif board[i][j] == 1 and (0 <= count < 2 or count > 3):
                    board[i][j] = 3

        for i in range(m):
            for j in range(n):
                if board[i][j] == 2:
                    board[i][j] = 1
                elif board[i][j] == 3:
                    board[i][j] = 0

File: algorithms/[289]生命游戏/solution.ts
// topics = ["数组"]

function gameOfLife(board: number[][]): void {
  const m = board.length;
  const n = board[0].length;

  for (let i = 0; i < m; i += 1) {
    for (let j = 0; j < n; j += 1) {
      board[i][j] = helper(board, i, j);
    }
  }

  for (let i = 0; i < m; i += 1) {
    for (let j = 0; j < n; j += 1) {
      if (board[i][j] === 2) {
        board[i][j] = 1;
      } else if (board[i][j] === 3) {
        board[i][j] = 0;
      }
    }
  }
}

function helper(board: number[][], i: number, j: number): number {
  const ori = board[i][j];
  const count = [
    board[i - 1]?.[j - 1],
    board[i - 1]?.[j],
    board[i - 1]?.[j + 1],
    board[i][j - 1],
    board[i][j + 1],
    board[i + 1]?.[j - 1],
    board[i + 1]?.[j],
    board[i + 1]?.[j + 1],
  ].filter(v => v == 1 || v == 3).length;

  if (ori == 0 && count == 3) return 2;
  if (ori == 1 && count != 2 && count != 3) return 3;
  return ori;
}
